### DISCUSSION TITLE
The use of a fast sample rate on ADS1115 in combination of the use  of GPIO pins on MCP2221

### DISCUSSION THREAD

**Discussion by Jur-I2C** [electronicayciencia/EasyMCP2221#15] [General]
@electronicayciencia : I saw in the EasyMCP library that you have recently added examples of python code that even collects data at rate of 167 Samples per second. Amazing ! 

I tried to use your code in an example which you can find below. In this example, I would like to blink a LED on GPIO pin 2 of the MCP2221 when the bit value of de ADC is below 10000. Unfortunately, I get an error. 

Here is the python code:

```python
import matplotlib.pyplot as plt
import time
import ADS1x15
import EasyMCP2221
from time import sleep

mcp = EasyMCP2221.Device()
mcp.set_pin_function(gp2 = "GPIO_OUT")

ADS = ADS1x15.ADS1115(1, 0x48)

#-------------config settings of ADS1115-----------
ADS.setGain(ADS.PGA_6_144V)
f = ADS.toVoltage()
ADS.setDataRate(ADS.DR_ADS111X_128)
ADS.setInput(ADS.INPUT_SINGLE_0)
ADS.setMode(ADS.MODE_CONTINUOUS) #--fastest mode for data aquisition

##-----------------------------
V = [ ] #--- define empty vectors
T = [ ]

ADS.requestADC(0) #----- first trigger, only read channel 0 form ADS1115
ADS.getValue()

start = time.time()
ntot=1000
for n in range (0,ntot):   
#val_0 = ADS.readADC(0)                                              #--- 33 Samples/ second (SPS) (method 1)

#val_0=ADS.getValue()                                                  #---83 SPS (method 2)

raw = ADS.i2c._read(0x48, 2)                                         #------167 SPS  (method 3)
val_0 = int.from_bytes(raw, "big", signed=True)

T.append(time.time() - start)                             #-- add data to vectors
V.append(val_0)

if val_0 < 10000 :        # let the LED blink on  GPIO pin 2 when val_0 is below 10000 
     mcp.GPIO_write(gp2 = True)
     sleep(0.05)
     mcp.GPIO_write(gp2 = False)
     sleep(0.05)

#print("Analog0: {0:d}\t{1:.3f} V".format(val_0, val_0 * f))   #---print values on screen (slowdowns the SPS)

end = time.time()
print(' Sample rate was : ' , ntot/(end - start) )

#--------------print the collected data in a graph-----------------
plt.plot(T, V, '-')
plt.xlabel("Time (s)")
plt.ylabel("Bit value ")
plt.title("A0 signal ")
plt.grid()
plt.show()
```

This is a quite straight forward program. But it does not work. I have the feeling that the I2C read-out sequence is disturbed when I use the GPIO pins of the MCP2221 breakout board. Anyway, I got the following error :

```console
Python 3.10.11 (tags/v3.10.11:7d4cc5a, Apr 5 2023, 00:38:17) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license()" for more information.

= RESTART: C:/Documents/PythonFiles/ReadoutTest_short.py
Traceback (most recent call last):
File "C://Documents/PythonFiles/ReadoutTest_short.py", line 32, in
raw = ADS.i2c._read(0x48, 2) #------167 SPS (method 3)
File "C:\Program Files\Python310\lib\site-packages\EasyMCP2221\smbus.py", line 45, in _read
return self.mcp.I2C_read(addr, length)
File "C:\Program Files\Python310\lib\site-packages\EasyMCP2221\MCP2221.py", line 1799, in I2C_read
raise RuntimeError("I2C read error. Internal status %02x." % (rbuf[I2C_INTERNAL_STATUS_BYTE]))
RuntimeError: I2C read error. Internal status 00.
```

Is there a solution for this problem or can I not use the SDA and SCL lines in combination with the GPIO pins on the MCP2221 breakout board simultaneously ?

**Comment by electronicayciencia** [comment]
This happens because you are driving the same MCP2221 chip with two instances of the EasyMCP2221 library. One from ADS1x15 and the other from your own `mcp = EasyMCP2221.Device()` line.

To prevent this you have two options:

- A. Rewrite the init of ADS1x15 library to allow external SMBus initialization. Like the example [Real Time Clock with LCD](https://easymcp2221.readthedocs.io/en/latest/smbus.html#example-2-real-time-clock-with-lcd).
- B. Reuse the `EasyMCP2221.Device` extracting it from the inner of the ADS1x15 library.

The easiest in this case is B.

```python
# No Need to import EasyMCP2221
import ADS1x15

# Initializing ADS1x15 will also initialize EasyMCP2221
ADS = ADS1x15.ADS1115(1, 0x48)

# Get the EasyMCP2221.Device object to reuse it
mcp = ADS.i2c.mcp

# Operate as usual
mcp.set_pin_function(gp2 = "GPIO_OUT")
```

Check the example [ads1115_with_gpio.py](https://github.com/electronicayciencia/EasyMCP2221/blob/master/examples/ADS1115/ads1115_with_gpio.py).

Is there a good reason you want to use an external I2C ADC? It may be easier for you to start with the internal MCP2221's ADC instead of an ADS1115. Check [level_meter.py](https://github.com/electronicayciencia/EasyMCP2221/blob/master/examples/level_meter.py).

**Reply by Jur-I2C** [reply]
The only reason why I use an external ADC is because of the limited resolution of the ADC on the MCP2221 breakout board.  It has only 10 bits resolution while the ADS1115 has a 16 bit resolution. The signal, I have to measure is in the range of -6 Volts to 6 Volts. 

Again thanks for writing an example in Python code.  I will implement this in my code.

**Comment by electronicayciencia** [comment]
I've just added a feature to prevent this double initialization. Not released yet. But feel free to checkout the repository and try it if you wish. Documented in [this hint](https://easymcp2221.readthedocs.io/en/latest/api_reference.html#device-initialization).
